Prepare Command API for migrating Order view page

sarjon opened the pull request:
<!--
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
 -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | This PR prepares new Command API for migrating "Order view" page.
| Type?         | refacto
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/10582
| How to test?  | Not yet.
<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/13712)
<!-- Reviewable:end -->

sarjon committed:
Adds AddProductToOrder command/handler interface

sarjon committed:
Adds part of implementation for adding product to exiting order

sarjon committed:
Finishes AddProductToOrder command/handler

sarjon committed:
Adds Behat tests for adding product to an existing order

sarjon committed:
Adds DeleteProductFromOrder command/handler

sarjon committed:
Adds UpdateOrderStatus command/handler

sarjon committed:
Adds UpdateProductInOrder command/handler

sarjon committed:
Adds ChangeOrderDeliveryAddress command/handler

sarjon committed:
Adds ChangeOrderInvoiceAddress command/handler

sarjon committed:
Adds ChangeOrderCurrency command/handler

sarjon committed:
Adds AddOrderPayment command/handler

sarjon commented:
@matks one more PR for you to review.

sarjon committed:
Adds GenerateOrderInvoice command/handler

sarjon committed:
Adds behat tests for generating order invoice

sarjon committed:
Adds UpdateOrderShippingDetails command/handler

matks commented on src/Adapter/Order/CommandHandler/AbstractOrderCommandHandler.php:
```diff
@@ -0,0 +1,174 @@
+    protected function reinjectQuantity(OrderDetail $orderDetail, $productQuantity, $delete = false)
+    {
+        // Reinject product
+        $reinjetableQuantity = (int) $orderDetail->product_quantity - (int) $orderDetail->product_quantity_reinjected;
```
```suggestion
        $reinjectableQuantity = (int) $orderDetail->product_quantity - (int) $orderDetail->product_quantity_reinjected;
```

matks replied:
```suggestion
        $reinjectableQuantity = (int) $orderDetail->product_quantity - (int) $orderDetail->product_quantity_reinjected;
```

matks commented on src/Adapter/Order/CommandHandler/AbstractOrderCommandHandler.php:
```diff
@@ -0,0 +1,174 @@
+     * @param int $productQuantity
+     * @param bool $delete
+     */
+    protected function reinjectQuantity(OrderDetail $orderDetail, $productQuantity, $delete = false)
```
Does this function really beong to an OrderCommandHandler ? It rather seems to be a generic Product / Stock processing.

Almost feels like a ProductCommandHandler or a ProductCommanderTrait 🤔 .

Again we are walking on the boundaries between domains 😅

sarjon replied:
Hmm, I think it's part of `Order` domain. It was used only in `AdminOrdersController` before.

> Does this function really beong to an OrderCommandHandler ? It rather seems to be a generic Product / Stock processing.

It restocks products **only from Order** by using `OrderDetail`, so it's related to `Orders` more or less I think.

matks replied:
👍

matks commented on src/Adapter/Order/CommandHandler/AddOrderPaymentHandler.php:
```diff
@@ -0,0 +1,80 @@
+    {
+        $order = $this->getOrderObject($command->getOrderId());
+
+        $amount = str_replace(',', '.', $command->getPaymentAmount());
```
😭

sarjon replied:
Should I change it? I'm a bit worried about changing legacy code, even this one. :/

matks replied:
No, we cannot change it, at least without an heavy test coverage for this usecase.
When we have working tests for this, we can then modify it and check nothing is broken.

matks commented on src/Adapter/Order/CommandHandler/AddProductToOrderHandler.php:
```diff
@@ -0,0 +1,506 @@
+        );
+
+        // update totals amount of order
+        $order->total_products += (float) $cart->getOrderTotal(false, Cart::ONLY_PRODUCTS);
```
All of this was being done in the legacy controller 😮 ?

sarjon replied:
Yup. :1st_place_medal:

matks commented on tests/Integration/Behaviour/Features/Scenario/Order/generate_invoice.feature:
```diff
@@ -0,0 +1,10 @@
+    Given there is order with reference "XKBKNABJK"
+    And order "XKBKNABJK" does not have any invoices
+    When I generate invoice for "XKBKNABJK" order
+    Then order "XKBKNABJK" should have invoice
```
Maybe we can assert a few properties from the invoice to check it was correctly generated ?

Also if there is a usecase where an invoice cannot be generated from the order (because of a bad status for example), we can add it here 😄

sarjon replied:
Sure, I have a few questions:
1. Is it ok to use demo data in behat tests?
2. Is this scenario ok? Do you have any more suggestions regarding steps?

matks replied:
1. Yes it's OK to use demo data in behat tests if this helps to avoid big setup steps like "I create a Country, I create a Customer, I create an address for this customer for this country, I create a Category, I create a product in this Category ..."

At some point we'll probably have to expand the demo data to have better usecases ready to be used. Like a visitor, a customer with 1 order, a customer with a lot of orders, a customer with a lot of carts...

2. This scenario would benefit from a little more assertions 😉we want to catch bad behaviors created from our modifications of the code. Comparing the invoice amount with the order amount could be interesting for example as it would help us feel safe when modifying the computation method.

matks replied:
🆙

sarjon committed:
Adds RemoveCartRuleFromOrder command/handler

sarjon committed:
Adds AddCartRuleToOrder command/handler

sarjon committed:
Fixes typo

sarjon committed:
Renames RemoveCartRuleFromOrder to DeleteCartRuleFromOrder

sarjon committed:
Adds UpdateInvoiceNote command/handler

sarjon commented on src/Core/Domain/Order/Invoice/Command/UpdateInvoiceNoteCommand.php:
```diff
@@ -0,0 +1,69 @@
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Core\Domain\Order\Invoice\Command;
```
@matks this is something I want to discuss with you. 

I've added `Invoice` subdomain as child for `Domain/Order` instead of putting it at the root (e.g. `Domain/Invoice`). At the moment it has only one `UpdateInvoiceNoteCommand` command, but I could move `GenerateOrderInvoiceCommand` (and rename to simply `GenerateInvoiceCommand`) under this subdomain as well. 

This kind of adds up for me, because Invoicing does not make sense without Orders, however, I can see that Invoicing could potentially live on it's own as well.

So, wdyt? Should I keep `Domain/Order/Invoice` or move it to `Domain/Invoice`? :thinking:

matks replied:
OK, this one is a tough one.

To answer it, I went to check the Orders > Invoices BO page. From what I see, I think that *for PrestaShop*, invoices without orders mean nothing.

There's no accounting lifecycle, there is no invoices payment cycle or management, I dont see any complex invoices actions.

Invoices seem to be only an order subdomain for now, so let's for for `Domain/Order/Invoice`

sarjon replied:
Ok, this makes sense for me too. :+1: 

However, there's more to it. Currently there a quite a few commands in `Domain/Order/Command` (there are more in other PR as well) that are not directly interacting with `Order`, but rather with order `Product` aka `OrderDetail` aka `OrderLine`. 

So I think that we could create another subdomain for `Order` which in my opinion makes sense since `Order` is quite a bit module/feature/whatever. It could be called `Domain/Order/OrderDetail` if we want to follow current naming or `Domain/Order/OrderLine` which is also a common name for it in ecommerce.

wdyt @matks?

matks replied:
Not sure about this one 🤔 but I guess if we dont do it, we're going to have a huuuuuge `Order` domain, right ?

sarjon replied:
Yeah, it's very likely.

matks replied:
Then it's a yes 👍

sarjon committed:
Adds IssuePartialRefund command/handler

sarjon committed:
Moves GenerateOrderInvoice under Order/Invoice namespace

sarjon committed:
Updates naming

sarjon committed:
Moves AddPayment to Order/Payment subdomain

sarjon committed:
Fixes namespace

sarjon committed:
Adds Product sumbdomain for Order

sarjon committed:
Applies CS fixer

sarjon committed:
Applies CS fixer on tests

sarjon commented:
@matks this one is ready for review, just don't forget that handlers contains extracted code from controllers, so there might be some weird stuff inside.

PierreRambaud commented on tests/Integration/Behaviour/Features/Context/ProductFeatureContext.php:
```diff
@@ -634,4 +634,16 @@ public function productWithNameProductisVirtual($productName)
+        $productId = Product::getIdByReference($productReference);
+
+        if (!$productId) {
+            throw new \RuntimeException(sprintf('Product with reference "%s" does not exist.', $productReference));
```
ProductNotFoundException?

sarjon replied:
Hmm, it's a test, would it be handled differently? :thinking:

PierreRambaud replied:
It can use Core Exception, no?

sarjon replied:
Same question, would it be handled differently? I mean, is there a reason not use native exceptions in behat tests? :thinking:

PierreRambaud replied:
Nope, as I know it will be handle exactly in the same way.
It's just to be consistent with the Core, and if we need to catch a specific exception during a Context.

sarjon replied:
Hmmmm, @matks wdyt? I saw native exceptions were used everywhere in behat contexts.

matks replied:
I have always used only RuntimeException in behat tests, because Exception seemed too generic 😅 basically we are writing assertions here, and we dont want to catch them. We want it to fail because it's intended to stop behat from working IMO

LouiseBonnard commented on src/Core/Grid/Definition/Factory/CategoryGridDefinitionFactory.php:
```diff
@@ -192,12 +193,13 @@ protected function getFilters()
-                    'attr' => [
-                        'placeholder' => $this->trans('Search ID', [], 'Admin.Actions'),
+                    'choices' => [
+                        $this->translator->trans('Amount') => 'amount',
```
In _Admin.Global_ please.

LouiseBonnard commented on src/Core/Grid/Definition/Factory/CategoryGridDefinitionFactory.php:
```diff
@@ -192,12 +193,13 @@ protected function getFilters()
-                        'placeholder' => $this->trans('Search ID', [], 'Admin.Actions'),
+                    'choices' => [
+                        $this->translator->trans('Amount') => 'amount',
+                        $this->translator->trans('Percent') => 'percent',
```
In _Admin.Global_ please.

LouiseBonnard commented:
One question though: do we agree that wordings introduced with `OrderException`are not translatable and therefore do not need domains?

sarjon commented:
> One question though: do we agree that wordings introduced with OrderExceptionare not translatable and therefore do not need domains?

Yes, messages for exceptions are not translatable and are not meant for end user, but rather for developers only.

sarjon commented on src/Adapter/Order/CommandHandler/UpdateProductInOrderHandler.php:
```diff
@@ -0,0 +1,247 @@
+        $this->assertProductCanBeUpdated($command, $orderDetail, $order, $orderInvoice);
+
+        // @todo: check if quantity can be array
+        // If multiple product_quantity, the order details concern a product customized
```
This todo will be checked during page migration.

sarjon committed:
Applies minor fixes

sarjon committed:
Applies CS fixer

sarjon commented:
ping @matks

sarjon commented:
@matks could you rerun build? I'm not sure if that was a random error.

sarjon commented:
ping @matks it's ready for review.

matks commented on src/Adapter/Order/CommandHandler/AddCartRuleToOrderHandler.php:
```diff
@@ -0,0 +1,281 @@
+
+        $res = true;
+        foreach ($cartRules as &$cartRule) {
+            $cartRuleObj = new CartRule();
```
I would like to improve this although this is not the scope of this PR.
Do you think we can at least add `@todo` to indicate areas of improvements ?

Here we could create a `private` function to create the proper CartRuleObject.

matks commented on src/Adapter/Order/CommandHandler/AddCartRuleToOrderHandler.php:
```diff
@@ -0,0 +1,281 @@
+
+        if ($res) {
+            foreach ($cartRules as $orderInvoiceId => $cartRule) {
+                // Create OrderCartRule
```
`@todo: use private method`

matks commented on src/Adapter/Order/CommandHandler/AddCartRuleToOrderHandler.php:
```diff
@@ -0,0 +1,281 @@
+               throw new OrderException('The discount type is invalid.');
+        }
+
+        $res = true;
```
```suggestion
        $result = true;
```

matks commented on src/Adapter/Order/CommandHandler/AddProductToOrderHandler.php:
```diff
@@ -0,0 +1,506 @@
+            $command->getOrderInvoiceId() ?: 0
+        );
+
+        // update totals amount of order
```
```suggestion
        // update totals amount of order
        // @todo: use https://github.com/PrestaShop/decimal for prices computations
```

matks commented on src/Adapter/Order/CommandHandler/AddProductToOrderHandler.php:
```diff
@@ -0,0 +1,506 @@
+        Hook::exec('actionOrderEdited', ['order' => $order]);
+
+        $oldCartRules = Context::getContext()->cart->getCartRules();
+        $newCartRules = Context::getContext()->cart->getCartRules();
```
😮 what's the difference between `$oldCartRules` and `$newCartRules` ?

sarjon replied:
I've moved `$oldCartRules` to the top since they had to be retrieved from an old cart (because new cart is created somewhere in between).

matks replied:
But now there's nothing in between 🤔

sarjon replied:
There will be after I push changes.

matks commented on src/Adapter/Order/CommandHandler/AddProductToOrderHandler.php:
```diff
@@ -0,0 +1,506 @@
+            $orderCartRule->value_tax_excl = $values['tax_excl'];
+            $orderCartRule->add();
+
+            $order->total_discounts += $orderCartRule->value;
```
```suggestion
            // @todo: use https://github.com/PrestaShop/decimal
            $order->total_discounts += $orderCartRule->value;
```

matks commented on src/Adapter/Order/CommandHandler/AddProductToOrderHandler.php:
```diff
@@ -0,0 +1,506 @@
+
+        $cart->add();
+
+        Context::getContext()->cart = $cart;
```
Created https://github.com/PrestaShop/PrestaShop/issues/14131 to solve this in the future

matks commented on src/Adapter/Order/CommandHandler/AddProductToOrderHandler.php:
```diff
@@ -0,0 +1,506 @@
+        );
+
+        if ($command->getProductPriceTaxIncluded() != $initialProductPriceTaxIncl) {
+            $specificPrice = new SpecificPrice();
```
`@todo: use private method to create specific price object`

matks commented on src/Adapter/Order/CommandHandler/AddProductToOrderHandler.php:
```diff
@@ -0,0 +1,506 @@
+
+        // Create Cart rule in order to make free shipping
+        if ($isFreeShipping) {
+            $freeShippingCartRule = new CartRule();
```
`@todo: use private method to create cart rule`

matks commented on src/Adapter/Order/CommandHandler/AddProductToOrderHandler.php:
```diff
@@ -0,0 +1,506 @@
+        $carrier = new Carrier((int) $order->id_carrier);
+        $taxCalculator = $carrier->getTaxCalculator($invoice_address);
+
+        $invoice->total_paid_tax_excl = Tools::ps_round((float) $cart->getOrderTotal(false, $totalMethod), 2);
```
`@todo: use https://github.com/PrestaShop/decimal to compute prices and taxes`

matks commented on src/Adapter/Order/CommandHandler/ChangeOrderCurrencyHandler.php:
```diff
@@ -0,0 +1,155 @@
+        foreach ($order->getOrderDetailList() as $orderDetail) {
+            $order_detail = new OrderDetail($orderDetail['id_order_detail']);
+
+            $fields = [
```
`@todo: use private method to handle this`

matks commented on src/Adapter/Order/CommandHandler/ChangeOrderCurrencyHandler.php:
```diff
@@ -0,0 +1,155 @@
+        }
+
+        // Update order && order_invoice amount
+        $fields = [
```
`@todo: use private method to handle this`

matks commented on src/Adapter/Order/CommandHandler/ChangeOrderDeliveryAddressHandler.php:
```diff
@@ -0,0 +1,63 @@
+        $order = $this->getOrderObject($command->getOrderId());
+        $address = new Address($command->getNewDeliveryAddressId()->getValue());
+
+        $cart = Cart::getCartByOrderId($order->id);
```
We modify both the cart and the order ? It's not one or the other ?

sarjon replied:
Yeah, seems we have to modify both, as they are different objects, but related.

matks commented:
I've got one good news and one bad news :trollface: 

The bad news: when I see all the processings you have extracted from the AdminOrdersController, I think "Woah ! So crazy !" then "Sarunas must hate me now" and finally "Sarunas is going to hate me even more".

The 3rd one is because *the code is extracted is a mess*. Yes, I know this is legacy code but we cannot let it like this.

I think we need at least to define a proper mapping of who does what.

For example, according to `DeleteCartRuleFromOrderHandler` implementation, to delete a cart rule from an order means:
- update the related invoice "tax" and "paid" fields
- update the related order "tax" and "paid" fields
- delete the cart rule from the order

I'd like us to modify the code to make it very clear. Like
```
public function handle(DeleteCartRuleFromOrderCommand $command)
    {
        $this->modifyRelatedInvoiceAsPriceEvolves($command);
        $this->modifyRelatedOrderAsPriceEvolves($command);

        $this->deleteCartRule($command);
```

Then we'll have 3 private functions that, later, we'll be able to plug on correct implementations.
So it'll start like this:
```
    private function modifyRelatedInvoiceAsPriceEvolves($command) {
            $orderInvoice = new OrderInvoice($orderCartRule->id_order_invoice);
            if (!Validate::isLoadedObject($orderInvoice)) {
                throw new OrderException('Can\'t load Order Invoice object');
            }
            // Update amounts of Order Invoice
            $orderInvoice->total_discount_tax_excl -= $orderCartRule->value_tax_excl;
            $orderInvoice->total_discount_tax_incl -= $orderCartRule->value;
            $orderInvoice->total_paid_tax_excl += $orderCartRule->value_tax_excl;
            $orderInvoice->total_paid_tax_incl += $orderCartRule->value;
            // Update Order Invoice
            $orderInvoice->update();
    }
```
but one day we'll change it to
```
    private function modifyRelatedInvoiceAsPriceEvolves($command) {
            $this->orderInvoiceManagementSystem->modifyWhenCartRuleIsDeleted($orderCartRule->id_order_invoice);
    }
```
To sum it up:
- I want to make the code easier to read using private function very obvious naming
- I want to prepare the future to make things right easier later

Oh, the good news ? We'll do it in another PR, not this one 😅

matks commented on src/Adapter/Order/CommandHandler/DeleteProductFromOrderHandler.php:
```diff
@@ -0,0 +1,157 @@
+    {
+        if ($orderDetail->id_order_invoice != 0) {
+            $order_invoice = new OrderInvoice($orderDetail->id_order_invoice);
+            $order_invoice->total_paid_tax_excl -= $orderDetail->total_price_tax_excl;
```
`@todo: use https://github.com/PrestaShop/decimal for price computations`

matks commented on src/Adapter/Order/CommandHandler/DeleteProductFromOrderHandler.php:
```diff
@@ -0,0 +1,157 @@
+     */
+    private function updateOrder(Order $order, OrderDetail $orderDetail)
+    {
+        $order->total_paid -= $orderDetail->total_price_tax_incl;
```
`@todo: use https://github.com/PrestaShop/decimal for price computations`

matks commented on src/Adapter/Order/CommandHandler/IssuePartialRefundHandler.php:
```diff
@@ -0,0 +1,293 @@
+            }
+        }
+
+        $shippingCostAmount = (float) str_replace(',', '.', $command->getShippingCostRefundAmount()) ?: false;
```
`@todo: use dedicated processing to deal with this issue (commas instead of colons)`

matks commented on src/Adapter/Order/CommandHandler/IssuePartialRefundHandler.php:
```diff
@@ -0,0 +1,293 @@
+        $voucher = 0;
+
+        if ($command->getCartRuleRefundType() === 1) {
+            $amount -= $voucher = (float) 99; //@todo: get order_discount_price
```
😐 ?

matks commented on src/Adapter/Order/CommandHandler/IssuePartialRefundHandler.php:
```diff
@@ -0,0 +1,293 @@
+            throw new OrderException('Please enter an amount to proceed with your refund.');
+        }
+
+        $choosen = false;
```
```suggestion
        $chosen = false;
```
?

sarjon replied:
I guess so. :man_shrugging:

matks commented on src/Adapter/Order/CommandHandler/IssuePartialRefundHandler.php:
```diff
@@ -0,0 +1,293 @@
+
+        if ($command->getCartRuleRefundType() === 1) {
+            $amount -= $voucher = (float) 99; //@todo: get order_discount_price
+        } elseif ($command->getCartRuleRefundType() === 2) {
```
Const to replace 2 ?

sarjon replied:
Yeah, but where do we put it? Should we have `Domain/Order/Refund` subdomain or should there be `Domain/Refund`? wdyt?

matks replied:
`Domain/Order/Refund` is OK because a refund always comes after an order has been checked out

matks commented on src/Adapter/Order/CommandHandler/IssuePartialRefundHandler.php:
```diff
@@ -0,0 +1,293 @@
+        if ($shippingCostAmount > 0) {
+            if (!$command->getTaxMethod()) {
+                $tax = new Tax();
+                $tax->rate = $order->carrier_tax_rate;
```
`@todo: use https://github.com/PrestaShop/decimal for price computations`

matks commented on src/Adapter/Order/CommandHandler/IssuePartialRefundHandler.php:
```diff
@@ -0,0 +1,293 @@
+            $translator = Context::getContext()->getTranslator();
+            $orderLanguage = new Language((int) $order->id_lang);
+
+            @Mail::Send(
```
`@todo: use a dedicated Mail class` (see https://github.com/PrestaShop/PrestaShop/issues/13945)

matks replied:
`@todo: remove this @ and have a proper error handling 😭 `

matks commented on src/Adapter/Order/CommandHandler/IssuePartialRefundHandler.php:
```diff
@@ -0,0 +1,293 @@
+
+            $customer = new Customer((int) $order->id_customer);
+
+            $params = [
```
`@todo: use private method`
like `sendMailWhatever()`

matks commented on src/Adapter/Order/CommandHandler/IssuePartialRefundHandler.php:
```diff
@@ -0,0 +1,293 @@
+            }
+
+            // Define a temporary code
+            $cartRule->code = sprintf('V0C%1$dO%2$d', $order->id_customer, $order->id);
```
`@todo: use private method`
and what is the reason for this temporary code 🤔 ?

sarjon replied:
I really don't know. It seems that this code is actually temporary as it is replaced later with new code that includes `$cartRule` id as well. Maybe this is some special format that is used somewhere, but it's my guess. :/

matks commented on src/Adapter/Order/CommandHandler/IssuePartialRefundHandler.php:
```diff
@@ -0,0 +1,293 @@
+                '{voucher_num}' => $cartRule->code,
+            ];
+
+            $orderLanguage = new Language((int) $order->id_lang);
```
`@todo: use private method to send mail and later a decoupled mail sender`

matks commented on src/Adapter/Order/CommandHandler/UpdateOrderStatusHandler.php:
```diff
@@ -0,0 +1,114 @@
+        }
+
+        // Save all changes
+        if ($history->addWithemail(true, $templateVars)) {
```
I almost did no see this because it's in the `if`. Can we get it out to make it more obvious ?

matks commented on src/Adapter/Order/CommandHandler/UpdateProductInOrderHandler.php:
```diff
@@ -0,0 +1,247 @@
+
+        $product_quantity = $command->getQuantity();
+
+        $product_price_tax_incl = Tools::ps_round(Tools::getValue('product_price_tax_incl'), 2);
```
`@todo: use https://github.com/PrestaShop/decimal for price computations`

matks commented on src/Adapter/Order/CommandHandler/UpdateProductInOrderHandler.php:
```diff
@@ -0,0 +1,247 @@
+        }
+
+        // Clean price
+        $product_price_tax_incl = str_replace(',', '.', $command->getPriceTaxIncluded());
```
Since the data comes from the Command, cannot we make sure it is correct and does need this sanitizing ?

sarjon replied:
I think no, because that would mean fixing issue quietly (as it is now here). I'll add todo for it so we can improve it, but what I think is that we should use Decimal or maybe Price VO that has validation inside.

matks commented on src/Adapter/Order/CommandHandler/UpdateProductInOrderHandler.php:
```diff
@@ -0,0 +1,247 @@
+            throw new OrderException('Invalid quantity');
+        }
+
+        // @todo: check if quantity can be array
```
We'll handle this later ?

sarjon replied:
Yes, during page migration.

matks commented on src/Core/Domain/Order/Command/AddCartRuleToOrderCommand.php:
```diff
@@ -0,0 +1,125 @@
+    private $orderInvoiceId;
+
+    /**
+     * @param int $orderId
```
phpDoc is outdated

matks commented on src/Core/Domain/Order/Command/AddCartRuleToOrderCommand.php:
```diff
@@ -0,0 +1,125 @@
+    }
+
+    /**
+     * @return float
```
I think we might start using types from https://github.com/PrestaShop/decimal here and cast them into float when they are passed to adapter handlers

sarjon replied:
Agree, but let's do it in separate PR.

matks commented on src/Core/Domain/Order/Command/DeleteCartRuleFromOrderCommand.php:
```diff
@@ -0,0 +1,71 @@
+    /**
+     * @var int
+     */
+    private $orderCartRule;
```
```suggestion
    private $orderCartRuleId;
```
?

matks commented on src/Core/Domain/Order/Command/IssuePartialRefundCommand.php:
```diff
@@ -0,0 +1,171 @@
+     */
+    public function __construct(
+        $orderId,
+        array $orderDetailRefunds,
```
What is in this array ? Can we assert better what it is, instead of just "array" ?

sarjon replied:
I prefer to do more specific assertions for commands when doing actual page migration, a lot of data types are not clear just yet and constructor arguments can also change in some cases.

At the moment this is unstructured data with inner items inside array.

matks commented on src/Core/Domain/Order/OrderDiscountType.php:
```diff
@@ -0,0 +1,48 @@
+/**
+ * Discount types that can be added to an order
+ */
+final class OrderDiscountType
```
👍

matks commented on src/Core/Domain/Order/Product/Command/AddProductToOrderCommand.php:
```diff
@@ -0,0 +1,236 @@
+    private $productId;
+
+    /**
+     * @var int
```
```suggestion
     * @var CombinationId
```
?

sarjon replied:
Yeah, where do we put it? Is it part of `Domain/Product/Combination`?

matks replied:
Yes 👍 looks right

matks commented on src/Core/Domain/Order/Product/Command/AddProductToOrderCommand.php:
```diff
@@ -0,0 +1,236 @@
+     *
+     * @return self
+     */
+    public static function toExistingInvoice(
```
👍

matks commented on src/Core/Domain/Order/Product/Command/DeleteProductFromOrderCommand.php:
```diff
@@ -0,0 +1,71 @@
+    private $orderId;
+
+    /**
+     * @var int
```
```suggestion
     * @var OrderDetailId
```
?

sarjon replied:
Same here, to what `Domain` does it belong? :thinking:

matks replied:
`Domain/Order` if the `domain/Order` folder is not too full yet
Else `Domain/Order/Detail`

matks commented on src/Core/Domain/Product/ValueObject/ProductId.php:
```diff
@@ -29,7 +29,7 @@
 
 /**
- * Product identity
+ * Product identity.
```
Nice :trollface:

sarjon replied:
Haha, this happened after rebase. :smile:

sarjon commented:
> I'd like us to modify the code to make it very clear. Like

That makes sense and I totally agree, however, in a lot of cases code is very coupled, for example, method is 300 lines of code, then variables create in the first 50 lines are used in the last 50 lines or somewhere in the middle, so it's not that easy to split the code into little private functions.

Of course it does not apply to every situation.

sarjon committed:
Address @matks review comments

sarjon commented:
@matks let's keep this PR as is (I've addresses some issues, but not all) and:
- [ ] Improve Behat tests with additional tests in separate PR
- [ ] Address refactoring of handlers in separate PR per handler (to keep PRs small)

wdyt?

matks commented:
@sarjon Agreed

matks commented:
@sarjon Behat tests are failing, which blocks Travis. Can you fix it, or do you want me to dive into it if the error is irrational ?

sarjon commented:
I'll check it, but it's weird since 5.6 passed, but 7.2 build failed. :man_shrugging:

matks commented:
> I'll check it, but it's weird since 5.6 passed, but 7.2 build failed. 🤷‍♂

You're right, it probably indicates a random failure 😞 

I re-run, we'll see

matks commented:
Interesting, it's not random.
```
--- Failed scenarios:
    Features/Scenario/Order/add_product.feature:7
234 scenarios (233 passed, 1 failed)
2821 steps (2819 passed, 1 failed, 1 skipped)
3m29.60s (55.27Mb)
Script @php -d date.timezone=UTC ./vendor/bin/behat -c tests/Integration/Behaviour/behat.yml handling the integration-behaviour-tests event returned with error code 1
```

sarjon commented:
> Interesting, it's not random.

Rerun again? Let's keep rerunning until it passes. :trollface:

matks commented:
I'm going to check locally with php7.2

matks committed:
Handle Add to Cart failure usecase where previous cart rules were empty or previous cart not available

matks commented:
Re-running Travis for e2e tests

sarjon commented:
All green, ready for merge?

matks commented:
Thank you @sarjon
